### PR TITLE
Feat/third party staking

### PR DESCRIPTION
**Pull request by franzns**
- Aura staking for pools
- Staking syncs now use new arch
- Aura user balances
- new user balance schema

### CODE REVIEW

File: modules/actions/pool/staking/sync-aura-staking.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            continue;
+        }
+
+        const stakingId = auraGauge.id + '-aura';
+
+        const dbStaking = pool.staking.find((staking) => staking.chain === chain && staking.id === stakingId);
+
+        if (!dbStaking) {
+            operations.push(
+                prisma.prismaPoolStaking.upsert({
```
Comment (gmbronco) [inline modules/actions/pool/staking/sync-aura-staking.ts:34]: sorry, i was here many times already, but always forget what's the "prismaPoolStaking", is it a gauge? and maybe we can simplify related tables, eg: "prismaPoolStakingAura"? I looks like they repeat a lot and one table with type and json can replace all 5 tables. eg:

```
incentives (or incentive_contracts)  {
  address, type, pool_id, chain, status, data (json – used to calculate the APR – stored separately?)
}
```

File: modules/actions/pool/staking/sync-aura-staking.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    auraSubgraphService: AuraSubgraphService,
+): Promise<void> => {
+    const auraGauges = await auraSubgraphService.getAllPools([chain]);
+    const pools = await prisma.prismaPool.findMany({
+        where: {
+            chain: chain,
+            address: { in: auraGauges.map((gauge) => gauge.lpToken.address) },
+        },
+        include: { staking: true },
+    });
```
Comment (gmbronco) [inline modules/actions/pool/staking/sync-aura-staking.ts:17]:
```suggestion
```

File: modules/actions/pool/staking/sync-aura-staking.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+        if (!pool) {
+            continue;
+        }
+
+        const stakingId = auraGauge.id + '-aura';
+
+        const dbStaking = pool.staking.find((staking) => staking.chain === chain && staking.id === stakingId);
+
+        if (!dbStaking) {
```
Comment (gmbronco) [inline modules/actions/pool/staking/sync-aura-staking.ts]:
```suggestion
```

Comment (gmbronco) [reply modules/actions/pool/staking/sync-aura-staking.ts]: how about always upserting what is in aura? and handling gracefully if we don't have a pool in a db? should be an edge case.

File: modules/actions/pool/staking/sync-aura-staking.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                create: {
+                    id: stakingId,
+                    chain: chain,
+                    stakingId: stakingId,
+                    apr: auraGauge.aprs.total,
+                    auraPoolAddress: auraGauge.address,
+                    isShutdown: auraGauge.isShutdown,
+                },
+                update: { apr: auraGauge.aprs.total },
+            }),
```
Comment (gmbronco) [inline modules/actions/pool/staking/sync-aura-staking.ts]: change to always upsert?

File: modules/actions/pool/staking/sync-aura-staking.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        );
+    }
+
+    await prismaBulkExecuteOperations(operations, true);
```
Comment (gmbronco) [inline modules/actions/pool/staking/sync-aura-staking.ts:65]: prismaBulkExecuteOperations uses promise.all, so the whole promise will get rejected. better to change to `.allSettled` - then we get the each of the results and can log an error when pool is missing, just to know when syncing is off.

**Review by gmbronco — COMMENTED** [review]
looks good! 2 points:

1. I'm not a fan of multiple tables for "stakings". Would take it as an opportunity to simplify the schema, your call.
2. When syncing data from Aura, you currently check if pools exist. Instead, you could let the database handle this by allowing it to reject an upsert if the foreign key check fails. This will simplify the process a bit.

**Comment by franzns** [comment]
Thought about simplifying the schema.
The only option I could come up with was using a jsonb column with staking type specific data, similar to what we do with the pool type data.

**Comment by gmbronco** [comment]
> The only option I could come up with was using a jsonb column with staking type specific data, similar to what we do with the pool type data.

good idea, i think it can work nicely.

**Pull request merged**